cmd/admin/build_instance.go

```diff
@@ -0,0 +1,203 @@
+		return fmt.Errorf("invalid time format, please use %s: %w", models.InstanceScheduledAtTimeLayout, err)
+	}
+
+	jobResponse, err := b.sendInstanceRequest(jobName, jobScheduledTimeProto)
+	if err != nil {
+		return fmt.Errorf("request failed for job %s: %w", jobName, err)
+	}
+	return b.writeInstanceResponse(jobResponse)
+}
+
+// writeInstanceResponse fetches a JobRun from the store (eg, postgres)
+// Based on the response, it builds assets like query, env and config
+// for the Job Run which is saved into output files.
+func (b *buildInstanceCommand) writeInstanceResponse(jobResponse *pb.RegisterInstanceResponse) (err error) {
+	dirPath := filepath.Join(b.assetOutputDir, taskInputDirectory)
+	if err := b.writeJobResponseMapToFiles(jobResponse, dirPath); err != nil {
+		return fmt.Errorf("error writing response map to file: %w", err)
+	}
+
+	if err := b.writeJobResponseEnvToFile(jobResponse, dirPath); err != nil {
+		return fmt.Errorf("error writing response env to file: %w", err)
+	}
+
+	if err := b.writeJobResponseSecretToFile(jobResponse, dirPath); err != nil {
+		return fmt.Errorf("error writing response env to file: %w", err)
+	}
+
+	if len(b.keysWithUnsubstitutedValue) > 0 {
+		b.logger.Warn(logger.ColoredNotice("Value not substituted for keys:\n%s", strings.Join(b.keysWithUnsubstitutedValue, "\n")))
+	}
+	return nil
+}
+
+func (b *buildInstanceCommand) writeJobResponseSecretToFile(
+	jobResponse *pb.RegisterInstanceResponse, dirPath string,
+) error {
+	// write all secrets into a file
+	secretsFileContent := ""
+	for key, val := range jobResponse.Context.Secrets {
+		if strings.Contains(val, unsubstitutedValue) {
+			b.keysWithUnsubstitutedValue = append(b.keysWithUnsubstitutedValue, key)
+		}
+		secretsFileContent += fmt.Sprintf("%s='%s'\n", key, val)
+	}
+
+	filePath := filepath.Join(dirPath, models.InstanceDataTypeSecretFileName)
+	writeToFileFn := utils.WriteStringToFileIndexed()
+	if err := writeToFileFn(filePath, secretsFileContent, b.logger.Writer()); err != nil {
+		return fmt.Errorf("failed to write asset file at %s: %w", filePath, err)
+	}
+	return nil
+}
+
+func (b *buildInstanceCommand) writeJobResponseEnvToFile(
+	jobResponse *pb.RegisterInstanceResponse, dirPath string,
+) error {
+	envFileBlob := ""
+	for key, val := range jobResponse.Context.Envs {
+		if strings.Contains(val, unsubstitutedValue) {
+			b.keysWithUnsubstitutedValue = append(b.keysWithUnsubstitutedValue, key)
+		}
+		envFileBlob += fmt.Sprintf("%s='%s'\n", key, val)
+	}
+
+	filePath := filepath.Join(dirPath, models.InstanceDataTypeEnvFileName)
+	writeToFileFn := utils.WriteStringToFileIndexed()
+	if err := writeToFileFn(filePath, envFileBlob, b.logger.Writer()); err != nil {
+		return fmt.Errorf("failed to write asset file at %s: %w", filePath, err)
+	}
+	return nil
+}
+
+func (b *buildInstanceCommand) writeJobResponseMapToFiles(
+	jobResponse *pb.RegisterInstanceResponse, dirPath string,
+) error {
+	if err := os.MkdirAll(dirPath, 0o777); err != nil {
```

Comment:
The file permission of `0o777` is too open, can we restrict it to `600` or similar

Reply:
updated